Enable existing service to be as `TransientService`

ghkim3221 opened the pull request:
### Motivation

 - Provide a way to make existing services as `TransientService`
   - In some cases, users want to make a service not to collect metrics like a `GrpcService` which implements [GRPC Health Checking Protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md)

### Modifications

 - Add `newDecorator` to `TransientHttpService` and `TransientRpcService`
 - Add `WrappingTransientHttpService` and `WrappingTransientRpcService` internally

### Result

 - Users can make existing service to be as `TransientService` by decorating with the `TransientHttpService` or `TransientRpcService` as follows:

   ```java
   builder.service("/docs", new DocService().decorate(
       TransientHttpService.newDecorator(TransientServiceOption.WITH_ACCESS_LOGGING)));
   ```

ikhoon commented:
How about hiding `WrappingTransientHttpService` and `WrappingTransientHttpService` from public API and adding factory methods to `TrasientService` and `TrasientRpcService`?

For example:
```java
public interface TrasientService {
    static Function<? super HttpClient, SimpleDecoratingHttpService> newDecorator(
            TransientServiceOption... transientServiceOptions) {
        ...
    }
}
```

ghkim3221 commented:
Thanks! nice idea.
In fact, I also added factory methods in `TransientHttpService` and `TransientRpcService`. But I realized that they are closed for extension if there are more properties in `TransientService`. So, I changed to public API with builder.
I'm not sure my thought is right. Please consider my intention and give me advice again. I also agree to your suggestion as I have done.

ikhoon commented:
> But I realized that they are closed for extension if there are more properties in TransientService.

Does that mean the factory method `newDecorator(TransientServiceOption... transientServiceOptions)` is not extendable? I think we add a builder method when we really need more options.

ghkim3221 commented:
Sorry for late. You right, I thought too far into the future.
Thanks @ikhoon and @trustin 👍, PTAL.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/TransientService.java:
```diff
@@ -31,6 +36,17 @@
+     * Returns a new {@link HttpService} decorator which makes the specified {@link HttpService} as
+     * {@link TransientService}.
+     */
+    static Function<? super HttpService, SimpleDecoratingHttpService> newDecorator(
```
Should we move this to `TransientHttpService`?

ikhoon replied:
Ah... you are right. I thought this is an `HttpService` like `LoggingService`. 😅

minwoox commented on core/src/main/java/com/linecorp/armeria/server/WrappingTransientHttpService.java:
```diff
@@ -0,0 +1,45 @@
+import com.linecorp.armeria.common.HttpResponse;
+
+/**
+ * Decorated a {@link HttpService} to be treated as {@link TransientService} without inheritance.
```
nit: Decorates an?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/WrappingTransientRpcService.java:
```diff
@@ -0,0 +1,45 @@
+import com.linecorp.armeria.common.RpcResponse;
+
+/**
+ * Decorated a {@link RpcService} to be treated as {@link TransientService} without inheritance.
```
ditto

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/TransientService.java:
```diff
@@ -31,6 +36,17 @@
+    static Function<? super HttpService, SimpleDecoratingHttpService> newDecorator(
+            TransientServiceOption... transientServiceOptions) {
+        requireNonNull(transientServiceOptions, "transientServiceOptions");
+        return delegate -> new WrappingTransientHttpService(delegate,
+                                                            ImmutableSet.copyOf(transientServiceOptions));
```
- Could use `EnumSet` such as `Sets.immutableEnumSet()`?
- Add another factory method for `Iterable<TransientServiceOption> transientServiceOptions`?

ikhoon commented on core/src/test/java/com/linecorp/armeria/server/WrappingTransientRpcServiceTest.java:
```diff
@@ -0,0 +1,45 @@
+
+class WrappingTransientRpcServiceTest {
+
+    static final RpcService FOO = (ctx, req) -> RpcResponse.of("foo");
```
nit: A `Service` is not constant and `FOO` is not good for a service name. Just use `fooService`?

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/TransientRpcService.java:
```diff
@@ -24,4 +30,15 @@
+        requireNonNull(transientServiceOptions, "transientServiceOptions");
+        return delegate -> new WrappingTransientRpcService(delegate,
+                                                           ImmutableSet.copyOf(transientServiceOptions));
+    }
```
Ditto
- Use `EnumSet`?
- Add a new method that takes `Iterable`.

ghkim3221 commented:
Thanks, @ikhoon and @minwoox! PTAL 😄

ikhoon reviewed (approved):
Thanks! @ghkim3221 🎄🎅🏻

trustin reviewed (approved):
Thanks, @ghkim3221 !

minwoox reviewed (approved):
Thanks, @ghkim3221!

trustin commented:
@ghkim3221 Would you mind updating the PR description so it matches what we have now?

ghkim3221 commented:
Sure

trustin commented:
Thanks, @ghkim3221. Merged :tada:
